go/runtime/bundle/manager.go

```diff
@@ -244,6 +317,89 @@ func (m *Manager) Add(path string) error {
 	return nil
 }
 
+// AddTemporary adds a bundle from the given temporary bundle file.
+func (m *Manager) AddTemporary(tmpName string, labels map[string]string) error {
+	path, err := m.deriveTemporaryBundlePath(tmpName, labels)
+	if err != nil {
+		return err
+	}
+
+	if err = m.Add(path, labels); err != nil {
+		return err
+	}
+
+	_ = os.Remove(path) // Cleanup temporary bundle after adding.
+	return nil
+}
+
+// Remove removes bundles matching the given labels.
+func (m *Manager) Remove(labels map[string]string) error {
+	if err := validateLabels(labels); err != nil {
+		return err
+	}
+
+	m.store.RemoveManifestsWithLabels(labels)
+	return nil
+}
+
+func (m *Manager) deriveTemporaryBundlePath(tmpName string, labels map[string]string) (string, error) {
+	tmpOriginPath, err := filepath.Abs(filepath.Join(m.tmpBundleDir, labels[LabelOrigin]))
+	if err != nil {
+		return "", err
+	}
+	tmpPath, err := filepath.Abs(filepath.Join(tmpOriginPath, tmpName))
+	if err != nil {
+		return "", err
+	}
+
+	// Ensure path does not traverse the origin directory as a sanity check.
+	if !strings.HasPrefix(strings.ToLower(tmpPath), strings.ToLower(tmpOriginPath)) {
+		return "", fmt.Errorf("malformed temporary bundle name")
+	}
+
+	return tmpPath, nil
+}
+
+// WriteTemporary writes the given data to a temporary file that can later be referenced as a
+// bundle.
+func (m *Manager) WriteTemporary(tmpName string, labels map[string]string, create bool, data []byte) error {
```

Comment:
This could be improved in 2 ways:
- The host defines the name of the file, thus removing the need for a temporary name. The name (which acts like a handle/identifier) is returned as a response. The caller can use append call to add more bytes. 
- Use manifest hash for the name. This way, we can optimize to download each bundle only once, but we can rewrite and explode it multiple times with different tag (which would be instance id or whatever).

Reply:
In the current design, the scheduler actually uses the instance ID as the temporary name, ensuring that multiple downloads from the same instance never generate more files. So I think it is a good idea that the app decides what the temporary name should be.

Reply:
Yes, that should be fine as well. I would just prefer to use manifest hash as a name, so that anyone can check if we already have this bundle. But this could be just premature optimization, so I'm resolving.